### ISSUE TITLE
Investigate System.Runtime.Notify refcounting

### ISSUE THREAD

**Issue by AnnaShaleva** [nspcc-dev/neo-go#3490]
Need to investigate how stackitem refcounting works for `System.Runtime.Notify` handler for both Go and C# nodes.

**Comment by AnnaShaleva** [comment]
For the record: I investigated NeoGo code and Notifications size is not restricted in any way. I also deployed a contract and tried to send a lot of notifications, so it's confirmed in practice that it's possible for NeoGo.

However, for C# node there's a chance that notifications number is restricted. The restriction goes from the way how C# VM refcounting works (hi, https://github.com/nspcc-dev/neo-go/issues/2500), and if emitted Notifications are refcounted properly in C#, then the behaviour of C# and Go nodes is different. But for now I can't tell for sure what is the behaviour of C# node, we need to check it on practice using the same contract as for Go node and check the resulting behaviour (but not in mainnet/testnet, please).

**Comment by AnnaShaleva** [comment]
OK, I can't find the branch, but the contract was based on `TestEngineLimits`.

**Comment by AnnaShaleva** [comment]
As I said earlier, Go node does not restrict the number of notifications, it's proven by#3627.

Happy news: C# node also allows to emit 2048+1 *large* notifications. I also tried with 2500 (tx was HALTed) and starting from 3000 notifications C# node does not responding (probably due to too slow `calculatenetworkfee`, but I'm sure that if I construct transaction manually and send the raw to C# node, then it will accept.
```
anna@kiwi:~/Documents/GitProjects/nspcc-dev/neo-go$ ./bin/neo-go contract invokefunction -w ./bench-wallet.json -r http://localhost:20331 05d5fd789874828b89f65e2ac842f9754386483e notify 5000
Enter account Nbbm89tWH2Adpuftd3perKb15KwExTTf1a password > 
Post "http://localhost:20331": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
anna@kiwi:~/Documents/GitProjects/nspcc-dev/neo-go$ ./bin/neo-go contract invokefunction -w ./bench-wallet.json -r http://localhost:20331 05d5fd789874828b89f65e2ac842f9754386483e notify 4000
Enter account Nbbm89tWH2Adpuftd3perKb15KwExTTf1a password > 
Post "http://localhost:20331": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
anna@kiwi:~/Documents/GitProjects/nspcc-dev/neo-go$ ./bin/neo-go contract invokefunction -w ./bench-wallet.json -r http://localhost:20331 05d5fd789874828b89f65e2ac842f9754386483e notify 3000
Enter account Nbbm89tWH2Adpuftd3perKb15KwExTTf1a password > 
JSON decoding: context deadline exceeded (Client.Timeout or context cancellation while reading body)
```

So here's the proof (I did it in NeoBench in order not to break testnet/mainnet):
* Contract code: [runtime.go.txt](https://github.com/user-attachments/files/17437207/runtime.go.txt)
* Contract config file: [runtime.yml.txt](https://github.com/user-attachments/files/17437213/runtime.yml.txt)
* The resulting application log (HALT state, all notificaitons are there) - it doesn't fit into GitHub limit, but believe me, it works.

And contract deploy/invocation steps:
```
anna@kiwi:~/Documents/GitProjects/nspcc-dev/neo-go$ ./bin/neo-go contract deploy -w ./bench-wallet.json -i ./examples/runtime/runtime.nef -m ./examples/runtime/runtime.manifest.json -r http://localhost:20331
Enter account Nbbm89tWH2Adpuftd3perKb15KwExTTf1a password > 
Network fee: 0.2111952
System fee: 19.9427577
Total fee: 20.1539529
Relay transaction (y|N)> y
98c1aa02f8ab25d2036976bad82fb1de2f694897162e67e21ce170b8461c7d82
Contract: 05d5fd789874828b89f65e2ac842f9754386483e
anna@kiwi:~/Documents/GitProjects/nspcc-dev/neo-go$ ./bin/neo-go contract invokefunction -w ./bench-wallet.json -r http://localhost:20331 05d5fd789874828b89f65e2ac842f9754386483e notify 1
Enter account Nbbm89tWH2Adpuftd3perKb15KwExTTf1a password > 
Network fee: 0.0118052
System fee: 0.0221361
Total fee: 0.0339413
Relay transaction (y|N)> y
92ff7b9612cc3cd746c05e28432fc353ea451e6f06cf48c8a485b678c504e7ee
anna@kiwi:~/Documents/GitProjects/nspcc-dev/neo-go$ ./bin/neo-go contract invokefunction -w ./bench-wallet.json -r http://localhost:20331 05d5fd789874828b89f65e2ac842f9754386483e notify 2049
Enter account Nbbm89tWH2Adpuftd3perKb15KwExTTf1a password > 
Network fee: 0.0118252
System fee: 20.1769137
Total fee: 20.1887389
Relay transaction (y|N)> y
fe111511ad9bd351cfe4645752d0c8b32ff0ee715536e48ee568e2c5fcd02c1a
```

Voilà, we can easily end up in broken mainnet, although it will cost a small fortune. We need restrictions.

**Comment by AnnaShaleva** [comment]
As a straightforward suggestion from my side: restrict the *number* of notifications directly, without refcounting:
```
anna@kiwi:~/Documents/GitProjects/nspcc-dev/neo-go$ git diff pkg/core/
diff --git a/pkg/core/interop/context.go b/pkg/core/interop/context.go
index d9ff39426..bc9c13e63 100644
--- a/pkg/core/interop/context.go
+++ b/pkg/core/interop/context.go
@@ -564,7 +564,10 @@ func (ic *Context) IsHardforkActivation(hf config.Hardfork) bool {
 }
 
 // AddNotification creates notification event and appends it to the notification list.
-func (ic *Context) AddNotification(hash util.Uint160, name string, item *stackitem.Array) {
+func (ic *Context) AddNotification(hash util.Uint160, name string, item *stackitem.Array) error {
+       if len(ic.Notifications) == vm.MaxStackSize {
+               return fmt.Errorf("max notifications count exceeded")
+       }
        ic.Notifications = append(ic.Notifications, state.NotificationEvent{
                ScriptHash: hash,
                Name:       name,
diff --git a/pkg/core/interop/runtime/engine.go b/pkg/core/interop/runtime/engine.go
index 3a16a4055..26bf8b410 100644
--- a/pkg/core/interop/runtime/engine.go
+++ b/pkg/core/interop/runtime/engine.go
@@ -114,7 +114,10 @@ func Notify(ic *interop.Context) error {
        if len(bytes) > MaxNotificationSize {
                return fmt.Errorf("notification size shouldn't exceed %d", MaxNotificationSize)
        }
-       ic.AddNotification(curHash, name, stackitem.DeepCopy(stackitem.NewArray(args), true).(*stackitem.Array))
+       err := ic.AddNotification(curHash, name, stackitem.DeepCopy(stackitem.NewArray(args), true).(*stackitem.Array))
+       if err != nil {
+               return err
+       }
        return nil
 }
```

We already have restrictions per a single serialized notification size:
https://github.com/nspcc-dev/neo-go/blob/8e1fdd5d70d9fa50ba50bc36cedff43ffc802141/pkg/core/interop/runtime/engine.go#L110-L116
so the combination of these two restrictions won't allow to DoS the node.

**Comment by roman-khimov** [comment]
Looking at the broader N3 picture, VM stack is temporary in its essence, it's just some space used for computation, while storage changes and notifications are side-effects. Storage changes are strictly permanent, we have no limitation on storage changes, but each write into it costs quite some GAS. Notifications are not permanent, not even required to be stored for correct chain operation, they're supposed to be much cheaper than storage and they are, but they're still mostly understood as stored/available at least for some time. In that sense tying them to VM limits (reusing the same counter) is not necessary, they can have limits of their own. Even though intuitively it looks like cheating, we can have more items in notifications than ever allowed by VM.

I'd try going with lower (than 2K) limit, we can analyze the chain data to see actual usage patterns.

**Comment by AnnaShaleva** [comment]
> we can analyze the chain data to see actual usage patterns.

Here's the summary of N3 mainnet data up to 6220598 height for Application execution trigger (transactions). The maximum number of events per transaction is 2K (I have transactions hashes if needed), but I'd say that we'll be fine with MaxNotificationsCount set to 256.

| The number of notifications per transaction | Frequency of occurrence (frequency of occurrence in percents) |
| --- | --- |
| 0 | 2258750 (40.5907%) |
| 1 | 1018319 (18.2996%) |
| 2 | 626233 (11.2537%) |
| 3 | 299316 (5.3788%) |
| 4 | 174043 (3.1276%) |
| 5 | 818782 (14.7139%) |
| 6 | 36654 (0.6587%) |
| 7 | 47891 (0.8606%) |
| 8 | 13738 (0.2469%) |
| 9 | 145925 (2.6223%) |
| 10 | 39524 (0.7103%) |
| 11 | 14519 (0.2609%) |
| 12 | 8713 (0.1566%) |
| 13 | 17652 (0.3172%) |
| 14 | 15105 (0.2714%) |
| 15 | 2127 (0.0382%) |
| 16 | 5800 (0.1042%) |
| 17 | 2296 (0.0413%) |
| 18 | 2144 (0.0385%) |
| 19 | 719 (0.0129%) |
| 20 | 3747 (0.0673%) |
| 21 | 571 (0.0103%) |
| 22 | 1126 (0.0202%) |
| 23 | 1118 (0.0201%) |
| 24 | 596 (0.0107%) |
| 25 | 370 (0.0066%) |
| 26 | 338 (0.0061%) |
| 27 | 84 (0.0015%) |
| 28 | 2566 (0.0461%) |
| 29 | 13 (0.0002%) |
| 30 | 364 (0.0065%) |
| 31 | 18 (0.0003%) |
| 32 | 664 (0.0119%) |
| 33 | 566 (0.0102%) |
| 34 | 49 (0.0009%) |
| 35 | 94 (0.0017%) |
| 36 | 112 (0.0020%) |
| 37 | 16 (0.0003%) |
| 38 | 442 (0.0079%) |
| 39 | 45 (0.0008%) |
| 40 | 136 (0.0024%) |
| 41 | 8 (0.0001%) |
| 42 | 579 (0.0104%) |
| 43 | 227 (0.0041%) |
| 44 | 50 (0.0009%) |
| 45 | 40 (0.0007%) |
| 46 | 15 (0.0003%) |
| 47 | 3 (0.0001%) |
| 48 | 174 (0.0031%) |
| 49 | 20 (0.0004%) |
| 50 | 94 (0.0017%) |
| 51 | 9 (0.0002%) |
| 52 | 32 (0.0006%) |
| 53 | 76 (0.0014%) |
| 54 | 36 (0.0006%) |
| 55 | 23 (0.0004%) |
| 56 | 231 (0.0042%) |
| 57 | 10 (0.0002%) |
| 58 | 37 (0.0007%) |
| 59 | 3 (0.0001%) |
| 60 | 41 (0.0007%) |
| 61 | 23 (0.0004%) |
| 62 | 16 (0.0003%) |
| 63 | 19 (0.0003%) |
| 64 | 390 (0.0070%) |
| 65 | 14 (0.0003%) |
| 66 | 15 (0.0003%) |
| 67 | 2 (0.0000%) |
| 68 | 24 (0.0004%) |
| 69 | 1 (0.0000%) |
| 70 | 147 (0.0026%) |
| 71 | 1 (0.0000%) |
| 72 | 24 (0.0004%) |
| 73 | 2 (0.0000%) |
| 74 | 6 (0.0001%) |
| 75 | 14 (0.0003%) |
| 76 | 10 (0.0002%) |
| 77 | 1 (0.0000%) |
| 78 | 9 (0.0002%) |
| 79 | 6 (0.0001%) |
| 80 | 16 (0.0003%) |
| 81 | 2 (0.0000%) |
| 82 | 3 (0.0001%) |
| 83 | 5 (0.0001%) |
| 84 | 113 (0.0020%) |
| 85 | 6 (0.0001%) |
| 86 | 10 (0.0002%) |
| 87 | 4 (0.0001%) |
| 88 | 7 (0.0001%) |
| 90 | 10 (0.0002%) |
| 91 | 5 (0.0001%) |
| 92 | 15 (0.0003%) |
| 93 | 4 (0.0001%) |
| 94 | 5 (0.0001%) |
| 95 | 3 (0.0001%) |
| 96 | 5 (0.0001%) |
| 97 | 2 (0.0000%) |
| 98 | 46 (0.0008%) |
| 99 | 3 (0.0001%) |
| 100 | 59 (0.0011%) |
| 101 | 3 (0.0001%) |
| 102 | 6 (0.0001%) |
| 103 | 1 (0.0000%) |
| 104 | 10 (0.0002%) |
| 105 | 3 (0.0001%) |
| 106 | 4 (0.0001%) |
| 107 | 1 (0.0000%) |
| 108 | 3 (0.0001%) |
| 109 | 1 (0.0000%) |
| 110 | 6 (0.0001%) |
| 112 | 20 (0.0004%) |
| 113 | 1 (0.0000%) |
| 114 | 2 (0.0000%) |
| 115 | 1 (0.0000%) |
| 116 | 5 (0.0001%) |
| 117 | 3 (0.0001%) |
| 118 | 3 (0.0001%) |
| 119 | 1 (0.0000%) |
| 120 | 5 (0.0001%) |
| 121 | 1 (0.0000%) |
| 122 | 4 (0.0001%) |
| 123 | 2 (0.0000%) |
| 124 | 3 (0.0001%) |
| 125 | 7 (0.0001%) |
| 126 | 15 (0.0003%) |
| 127 | 1 (0.0000%) |
| 128 | 195 (0.0035%) |
| 129 | 2 (0.0000%) |
| 130 | 4 (0.0001%) |
| 131 | 2 (0.0000%) |
| 132 | 7 (0.0001%) |
| 133 | 1 (0.0000%) |
| 134 | 2 (0.0000%) |
| 135 | 4 (0.0001%) |
| 136 | 4 (0.0001%) |
| 137 | 1 (0.0000%) |
| 139 | 1 (0.0000%) |
| 140 | 11 (0.0002%) |
| 141 | 2 (0.0000%) |
| 142 | 1 (0.0000%) |
| 143 | 2 (0.0000%) |
| 144 | 6 (0.0001%) |
| 145 | 2 (0.0000%) |
| 147 | 3 (0.0001%) |
| 149 | 1 (0.0000%) |
| 150 | 3 (0.0001%) |
| 151 | 3 (0.0001%) |
| 154 | 12 (0.0002%) |
| 155 | 2 (0.0000%) |
| 156 | 2 (0.0000%) |
| 157 | 1 (0.0000%) |
| 158 | 1 (0.0000%) |
| 160 | 2 (0.0000%) |
| 165 | 1 (0.0000%) |
| 166 | 1 (0.0000%) |
| 167 | 1 (0.0000%) |
| 168 | 20 (0.0004%) |
| 171 | 4 (0.0001%) |
| 172 | 4 (0.0001%) |
| 174 | 4 (0.0001%) |
| 176 | 1 (0.0000%) |
| 177 | 1 (0.0000%) |
| 178 | 1 (0.0000%) |
| 182 | 9 (0.0002%) |
| 183 | 1 (0.0000%) |
| 184 | 2 (0.0000%) |
| 185 | 5 (0.0001%) |
| 186 | 5 (0.0001%) |
| 189 | 2 (0.0000%) |
| 190 | 2 (0.0000%) |
| 192 | 1 (0.0000%) |
| 194 | 1 (0.0000%) |
| 195 | 1 (0.0000%) |
| 196 | 1 (0.0000%) |
| 198 | 7 (0.0001%) |
| 199 | 2 (0.0000%) |
| 200 | 27 (0.0005%) |
| 201 | 2 (0.0000%) |
| 202 | 2 (0.0000%) |
| 203 | 1 (0.0000%) |
| 204 | 2 (0.0000%) |
| 205 | 1 (0.0000%) |
| 206 | 1 (0.0000%) |
| 208 | 1 (0.0000%) |
| 213 | 2 (0.0000%) |
| 214 | 16 (0.0003%) |
| 215 | 1 (0.0000%) |
| 216 | 4 (0.0001%) |
| 218 | 1 (0.0000%) |
| 220 | 1 (0.0000%) |
| 225 | 1 (0.0000%) |
| 228 | 10 (0.0002%) |
| 235 | 1 (0.0000%) |
| 237 | 1 (0.0000%) |
| 242 | 3 (0.0001%) |
| 243 | 1 (0.0000%) |
| 248 | 3 (0.0001%) |
| 249 | 1 (0.0000%) |
| 250 | 7 (0.0001%) |
| 256 | 1 (0.0000%) |
| 257 | 1 (0.0000%) |
| 260 | 3 (0.0001%) |
| 261 | 2 (0.0000%) |
| 262 | 1 (0.0000%) |
| 264 | 3 (0.0001%) |
| 266 | 1 (0.0000%) |
| 268 | 1 (0.0000%) |
| 270 | 2 (0.0000%) |
| 271 | 2 (0.0000%) |
| 272 | 1 (0.0000%) |
| 273 | 2 (0.0000%) |
| 276 | 1 (0.0000%) |
| 283 | 1 (0.0000%) |
| 284 | 1 (0.0000%) |
| 286 | 1 (0.0000%) |
| 288 | 1 (0.0000%) |
| 290 | 1 (0.0000%) |
| 291 | 1 (0.0000%) |
| 299 | 1 (0.0000%) |
| 300 | 6 (0.0001%) |
| 306 | 2 (0.0000%) |
| 307 | 1 (0.0000%) |
| 308 | 1 (0.0000%) |
| 309 | 1 (0.0000%) |
| 310 | 3 (0.0001%) |
| 311 | 1 (0.0000%) |
| 321 | 1 (0.0000%) |
| 329 | 1 (0.0000%) |
| 332 | 1 (0.0000%) |
| 333 | 1 (0.0000%) |
| 335 | 1 (0.0000%) |
| 342 | 2 (0.0000%) |
| 349 | 1 (0.0000%) |
| 350 | 1 (0.0000%) |
| 360 | 1 (0.0000%) |
| 378 | 1 (0.0000%) |
| 400 | 1 (0.0000%) |
| 406 | 1 (0.0000%) |
| 411 | 1 (0.0000%) |
| 420 | 1 (0.0000%) |
| 421 | 1 (0.0000%) |
| 422 | 1 (0.0000%) |
| 423 | 3 (0.0001%) |
| 430 | 1 (0.0000%) |
| 434 | 1 (0.0000%) |
| 437 | 1 (0.0000%) |
| 468 | 1 (0.0000%) |
| 496 | 1 (0.0000%) |
| 498 | 4 (0.0001%) |
| 499 | 4 (0.0001%) |
| 500 | 9 (0.0002%) |
| 501 | 1 (0.0000%) |
| 504 | 1 (0.0000%) |
| 513 | 1 (0.0000%) |
| 533 | 1 (0.0000%) |
| 539 | 1 (0.0000%) |
| 545 | 1 (0.0000%) |
| 558 | 1 (0.0000%) |
| 615 | 1 (0.0000%) |
| 620 | 1 (0.0000%) |
| 741 | 1 (0.0000%) |
| 836 | 1 (0.0000%) |
| 844 | 1 (0.0000%) |
| 924 | 1 (0.0000%) |
| 1000 | 1 (0.0000%) |
| 2000 | 4 (0.0001%) |

**Comment by roman-khimov** [comment]
Whoa. That's a _huge_ tail. And what are those 1000-2000 notification transactions? Are they real? Look like pure tests to me.

**Comment by roman-khimov** [comment]
> 28       PUSHDATA1    61697264726f70 ("airdrop")

So it's not a test, it's a token giveaway which can legitimately have a lot of transfers. I'd suggest 512 for the limit then, it should be sufficient for events like this (4 transactions are OK here) while restrictive enough to prevent problems in evil scenarios.

**Comment by roman-khimov** [comment]
Just for some reference:
https://ethereum.stackexchange.com/questions/1686/ethereum-event-log-maximum-size
https://ethereum.stackexchange.com/questions/99910/whats-the-maximum-number-of-logs-per-transaction

Our event (notification) price is _not_ dependent on size, maybe it's a part of the problem and we can solve it economically. Events are limited to 1024 bytes, but some regular transfer is ~50 bytes (20+20+int+overhead), so 20 regular events have the same volume as one big event, but you pay X for a biggy with 20X for regular transfers.

**Comment by roman-khimov** [comment]
Does C# node have any problems with 2000-5000 small notifications? I'm trying to understand what matters most, overall size or just the count.

**Comment by AnnaShaleva** [comment]
> Does C# node have any problems with 2000-5000 small notifications?

First of all, it should be noted that it's not proven that C# node has problems with *large* notifications. Everything I learnt is that C# RPC server has problems with such transaction construction:

> starting from 3000 notifications C# node does not responding (probably due to too slow calculatenetworkfee), but I'm sure that if I construct transaction manually and send the raw to C# node, then it will be accepted.

So I guess we firstly need to know how Go/C# node handle a pure `sendrawtransaction` request with manually constructed transaction. Regarding Go node: my laptop was just shutted down when I tried to construct transaction with 5000 notifications, it looked like OOM, and it should be noted that I had 4-nodes privnet running, so it will probably be better with a single node. Regarding C# node: the problem with RPC request timeout prevents me from constructing larger transaction, so I need to try to construct it manually and see if it's accepted to chain.

So let me try a standalone `sendrawtransaction` request, and after that I will try small notifications.

> I'm trying to understand what matters most, overall size or just the count.

From what I see and what I guess, the overall size of all resulting notifications matters, because this problem looks to me more like a potential OOM reason. So 5K of small objects may not affect the node if the overall objects size is not as big as 2K of large objects. But let's check.

**Comment by AnnaShaleva** [comment]
I constructed transaction manually and used RPC client only for `sendrawtransaction` request. I also added one more method to the contract that produces some small notification like, so the resulting methods look like:
```go
func NotifyLarge(count int) {
	for i := 0; i < count; i++ {
		runtime.Notify("LargeEvent", args...)
	}
}

func NotifySmall(count int) {
	for i := 0; i < count; i++ {
		runtime.Notify("SmallEvent", true)
	}
}
```
Here are results for NeoGo node:
* It perfectly handles 50K small notifications without any damage to the node or RPC server. Experiments with larger number of notifications were not conducted, I doubt we need them.
* It perfectly handles up to 10K large notifications without any damage to the node or RPC server. However, with 50K large notifications transaction is being accepted to the chain with a delay (~1.4s) in subsequent block acceptance:
```
2024/10/23 12:27:27 empty block: 75
2024/10/23 12:27:27 #75: 0 transactions in 5004 ms - 0.000000 tps
2024/10/23 12:27:28 CPU: 0.548%, Mem: 1187.469MB
2024/10/23 12:27:38 CPU: 42.023%, Mem: 3131.305MB
2024/10/23 12:27:42 #76: 1 transactions in 5016 ms - 0.199362 tps       <-- transaction with 50K notifications
2024/10/23 12:27:42 empty block: 77
2024/10/23 12:27:42 #77: 0 transactions in 6456 ms - 0.000000 tps       <-- next block is delayed [likely] due to previous block processing
2024/10/23 12:27:47 empty block: 78
2024/10/23 12:27:47 #78: 0 transactions in 8573 ms - 0.000000 tps       <-- next block is also delayed
...
// the rest of blocks are produced without delay
```
and RPC server can't handle `getapplicationlog` request for this transaction which results in RPC server shutdown:
```
2024/10/23 12:28:06 could not fetch block count: error after calling rpc server the server closed connection before returning the first response byte. Make sure the server returns 'Connection: close' response header before closing the connection
2024/10/23 12:28:06 could not fetch block count: error after calling rpc server error when dialing 172.200.0.6:20331: dial tcp4 172.200.0.6:20331: connect: connection refused
```
And here are the logs from the client side:
```
    client_test.go:189: 
        	Error Trace:	/home/anna/Documents/GitProjects/nspcc-dev/neo-go/pkg/services/rpcsrv/client_test.go:189
        	Error:      	Received unexpected error:
        	            	Post "http://localhost:20331": EOF
        	Test:       	TestClientXX
```
But CNs continue normal blocks accepting afterwards which is OK.

The results for C# node:
* It also perfectly handles 50K small notifications without any damage to the node or RPC server, everything works fine.
* Everything is OK with 2K large notifications. Transaction with 10K large notifications is also accepted to the chain (and RPC server serves `getapplicationlog` properly), but it takes significant amount of time to process this transaction and next block is delayed (on ~24 seconds). But consensus recovers afterwards:
```
2024/10/23 11:47:07 #35: 0 transactions in 5066 ms - 0.000000 tps
2024/10/23 11:47:08 CPU: 6.883%, Mem: 935.297MB
2024/10/23 11:47:12 empty block: 36
2024/10/23 11:47:12 #36: 0 transactions in 5045 ms - 0.000000 tps
2024/10/23 11:47:17 empty block: 37
2024/10/23 11:47:17 #37: 0 transactions in 5060 ms - 0.000000 tps
2024/10/23 11:47:18 CPU: 5.198%, Mem: 937.309MB
2024/10/23 11:47:28 CPU: 56.139%, Mem: 1621.500MB
2024/10/23 11:47:38 CPU: 53.563%, Mem: 3996.613MB
2024/10/23 11:47:49 CPU: 58.552%, Mem: 4744.695MB
2024/10/23 11:47:59 CPU: 39.740%, Mem: 5384.199MB
2024/10/23 11:47:59 #38: 1 transactions in 5047 ms - 0.198138 tps        <-- transaction with 10K notificaitons
2024/10/23 11:48:09 CPU: 16.929%, Mem: 5063.910MB
2024/10/23 11:48:13 empty block: 39
2024/10/23 11:48:13 #39: 0 transactions in 29656 ms - 0.000000 tps      <-- next block is delayed
2024/10/23 11:48:13 empty block: 40
2024/10/23 11:48:13 #40: 0 transactions in 7555 ms - 0.000000 tps        <-- next block is also delayed
...
// consensus recovers in ~2 blocks afterwards and continues with 5-second blocks.
```
* An attempt to send transaction with 50K large notifications killed my laptop, no block was accepted until that moment. But it looks to me like a matter of RAM available because I have four nodes running on my laptop, so things will probably get better if the node is running with more RAM available.

To summarize, I'm still stick to the previously added opinion: overall notifications size matters, not the number of notifications.

**Comment by AnnaShaleva** [comment]
Just for the record: here's some statistics for OnPersist/PostPersist triggers:
**OnPersist:**
```
anna@kiwi:~/Documents/GitProjects/nspcc-dev/neo-go$ tail stat_onPersist_sorted_2.txt 
2024-10-18T20:54:50.914+0300	INFO	NTF	{"trigger": "OnPersist", "height": 4555186, "container": "3b03f0b434c8b535b3a33cb2c811fc4efc8c602ca54f4b2d6578490b797ff843", "count": 513
2024-10-18T20:54:50.920+0300	INFO	NTF	{"trigger": "OnPersist", "height": 4555187, "container": "e8c2bbab347f89c5752065b9f1b34364a15b47a699d890189d6d6b5d4bb17943", "count": 513
2024-10-18T20:54:50.929+0300	INFO	NTF	{"trigger": "OnPersist", "height": 4555188, "container": "631bb32e586763abe630ffad4c997e1675f380cd19ab731ecbe7b971015d9f19", "count": 513
2024-10-18T19:29:06.475+0300	INFO	NTF	{"trigger": "OnPersist", "height": 1551543, "container": "19d66b329f6c65aacdb0f6b61b76a8c8f45410d7c27b9cbca5e9c39629983614", "count": 514
2024-10-18T19:34:33.366+0300	INFO	NTF	{"trigger": "OnPersist", "height": 1696758, "container": "19612c76a068e6d34947648827d7a5cbd34edf84a3456697e292d465a139d138", "count": 514
2024-10-18T20:53:32.385+0300	INFO	NTF	{"trigger": "OnPersist", "height": 4548957, "container": "10419d51bc2559dfabd861bdf4602449982dd9428dac25b47c3e0dbd3411323d", "count": 514
2024-10-18T20:53:46.403+0300	INFO	NTF	{"trigger": "OnPersist", "height": 4551834, "container": "0062d48b2b4de778ae1e66d4783864cb40dac0cc049c2b0cf0573c16959e2567", "count": 514
2024-10-18T20:54:34.630+0300	INFO	NTF	{"trigger": "OnPersist", "height": 4552422, "container": "bc6acc143e7bd3e3cf6cb62df37541f2460146e0f31528e5babb4340388e61d6", "count": 514
2024-10-18T20:54:48.222+0300	INFO	NTF	{"trigger": "OnPersist", "height": 4554669, "container": "717b7f7017a37d19460311781c812f72ba35ffb03751a48980af2acbd751be41", "count": 514
2024-10-18T20:54:49.030+0300	INFO	NTF	{"trigger": "OnPersist", "height": 4554795, "container": "3a5bc02d85f01551b4e7551bbf5b37ff96be9c5314f117541459847577be61ad", "count": 514
```
**PostPersist:**
```
anna@kiwi:~/Documents/GitProjects/nspcc-dev/neo-go$ tail stat_postPersist_sorted_2.txt 
2024-10-18T19:55:29.690+0300	INFO	NTF	{"trigger": "PostPersist", "height": 2506846, "container": "0024a8150fac70080497e9e6aa62ca60625cb4503d74fde45e1c6318c3287a11", "count": 5
2024-10-18T19:55:29.881+0300	INFO	NTF	{"trigger": "PostPersist", "height": 2506958, "container": "ae8561633648af9240c7461d37cdb307ded2f5903781590d92eb340921969f08", "count": 5
2024-10-18T19:55:30.054+0300	INFO	NTF	{"trigger": "PostPersist", "height": 2507192, "container": "68658a1690ad75fa0c8a23e1999ff8f3384621249172412f3a9fd3dcce5c3f85", "count": 5
2024-10-18T19:55:30.056+0300	INFO	NTF	{"trigger": "PostPersist", "height": 2507193, "container": "a2e8d4a5425f1b1e1d3d1e8a370f0810913c8db2e8813728e66398c5ab9eb90f", "count": 5
2024-10-18T19:56:13.316+0300	INFO	NTF	{"trigger": "PostPersist", "height": 2537763, "container": "268914054959e41e078a293334cdbc2e3ae5e396a16c63b2f46e0f43fefa948a", "count": 5
2024-10-18T19:56:13.329+0300	INFO	NTF	{"trigger": "PostPersist", "height": 2537767, "container": "f41a92faae78f28c0f5a1e8f870842747cfbf2cf50d80198484eac7435e4611d", "count": 5
2024-10-18T19:56:13.339+0300	INFO	NTF	{"trigger": "PostPersist", "height": 2537770, "container": "2de3aa5c3196a868e805a0a54321fbb1773caa8018f41549e25f8468d2a0e342", "count": 5
2024-10-18T19:56:13.347+0300	INFO	NTF	{"trigger": "PostPersist", "height": 2537773, "container": "b9edf56add9129112403893bd6c69e6f37a3edbdd03854aec67812de61e7ffa2", "count": 5
2024-10-18T19:57:39.317+0300	INFO	NTF	{"trigger": "PostPersist", "height": 2610383, "container": "05f5f31c48f01e0b3892ee8802d4368c4abaecd9dd6fc0ec053e2e3b4677c6cc", "count": 5
2024-10-18T19:57:39.354+0300	INFO	NTF	{"trigger": "PostPersist", "height": 2610463, "container": "b1be50ebeeedc4fbcec2523c57220e7f92461d87e47892c651936bdace14cd66", "count": 5
```

**Comment by roman-khimov** [comment]
It just follows the number of transactions (burning fees + sending something to the validator).
